File: Dynamic Programming I/P139.cpp
// 139. Word Break
// Given a string s and a dictionary of strings wordDict, return true if s can be segmented into a space-separated sequence of one or more dictionary words.
// Note that the same word in the dictionary may be reused multiple times in the segmentation.

// Example 1:
// Input: s = "leetcode", wordDict = ["leet","code"]
// Output: true
// Explanation: Return true because "leetcode" can be segmented as "leet code".

// Example 2:
// Input: s = "applepenapple", wordDict = ["apple","pen"]
// Output: true
// Explanation: Return true because "applepenapple" can be segmented as "apple pen apple".
// Note that you are allowed to reuse a dictionary word.

// Example 3:
// Input: s = "catsandog", wordDict = ["cats","dog","sand","and","cat"]
// Output: false

// Constraints:
// 1 <= s.length <= 300
// 1 <= wordDict.length <= 1000
// 1 <= wordDict[i].length <= 20
// s and wordDict[i] consist of only lowercase English letters.
// All the strings of wordDict are unique.

bool isContain(string str, const vector<string>& wordDict) {
	if (str.empty())
		return true;
	for (string word: wordDict) {
		if (word.compare(str) == 0)
			return true;
	}
	return false;
}

// dynamic programming approach
// time complexity O(n^2) where n is the length of string (assume searching the word in the dictionary takes O(1) time complexity)
// space complexity O(n)
bool wordBreak(string s, vector<string>& wordDict) {
	int n = s.length();
	vector<bool> dp(n + 1);
	dp[n] = true;
	dp[n - 1] = isContain(s.substr(n - 1, 1), wordDict);
	for (int i = n - 2; i >= 0; i--) {
		for (int j = i + 1; j <= n; j++) {
			string prefix = s.substr(i, j - i);
			if (isContain(prefix, wordDict) && dp[j]) {
				dp[i] = true;
				break;
			}
		}
	}
	return dp[0];
}

File: Algorithm-II/6-Dynamic-Programming/P139.java
package algorithm2.dynamicProgramming;

// date: Dec 23, 2021
// 139. Word Break
// https://leetcode.com/problems/word-break/

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class P139 {
    // dynamic programming approach
    // time complexity: O(n*n)
    // space complexity: O(n)
    public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        boolean[] dp = new boolean[n];
        Set<String> wordSet = new HashSet<>(wordDict);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i + 1; j++) {
                if ((j == 0 || dp[j - 1]) && wordSet.contains(s.substring(j, i + 1))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n - 1];
    }

    public static void main(String[] args) {
        P139 p = new P139();
        System.out.println("---------------------");
        System.out.println("Input: s = 'leetcode', wordDict = ['leet', 'code']. Expect: true");
        System.out.println("Output: " + p.wordBreak("leetcode", Arrays.asList("leet", "code")));
        System.out.println("---------------------");
        System.out.println("Input: s = 'applepenapple', wordDict = ['apple', 'pen']. Expect: true");
        System.out.println("Output: " + p.wordBreak("applepenapple", Arrays.asList("apple", "pen")));
        System.out.println("---------------------");
        System.out.println("Input: s = 'catsandog', wordDict = ['cats', 'dog', 'sand', 'and', 'cat']. Expect: false");
        System.out.println("Output: " + p.wordBreak("catsandog", Arrays.asList("cats", "dog", "sand", "and", "cat")));
    }
}

// reference
// https://leetcode.com/problems/word-break/discuss/44054/Java-DP-solution